data_source.go

```diff
@@ -171,27 +249,28 @@ type AnySource struct {
 	lastread     time.Time
 	nextFrameNum FrameIndex // frame number for the next frame we will receive
 	processors   []*DataStreamProcessor
-	abortSelf    chan struct{} // This can signal the Run() goroutine to stop
+	abortSelf    chan struct{}   // Signal to the core loop of active sources to stop
+	nextBlock    chan *dataBlock // Signal from the core loop that a block is ready to process
 	broker       *TriggerBroker
-	// publishSync  *PublishSync
+
 	shouldAutoRestart   bool // used to tell SourceControl to try to restart this source after an error
 	noProcess           bool // Set true only for testing.
 	heartbeats          chan Heartbeat
-	segments            []DataSegment
 	writingState        WritingState
 	numberWrittenTicker *time.Ticker
-	runMutex            sync.Mutex
+	sourceState         SourceState
+	runMutex            sync.Mutex // guards sourceState
 	runDone             sync.WaitGroup
 	readCounter         int
 }
 
-// ProcessSegments processes a single outstanding for each processor in ds
-// returns when all segments have been processed
-// it's a more synchnous version of each dsp launching it's own goroutine
-func (ds *AnySource) ProcessSegments() error {
+// ProcessSegments processes a single outstanding segment for each of ds.processors
+// Returns when all segments have been processed
+// It's a more synchronous version of each dsp launching its own goroutine
+func (ds *AnySource) ProcessSegments(block *dataBlock) error {
 	var wg sync.WaitGroup
 	for i, dsp := range ds.processors {
-		segment := ds.segments[i]
+		segment := block.segments[i]
 		if segment.processed {
```

Comment:
This can be removed I think. It only existed to check that I never started processing the same segments twice when I had `ds.segments` as a persistent variable.

Reply:
Okay, I'll remove the block.